### ISSUE TITLE
Cannot compile tests with clang 3.4.1: uint8_t vs. char

### ISSUE THREAD

**Issue by gperciva** [Tarsnap/scrypt#12]
So `crypto_scrypt()` is careful to always use `uint8_t`, whereas `test/main.c` uses `char`.

I can sprinkle a bunch of `(uint_t*)` in `test/main.c`, but then `strlen()` isn't happy with the result.  Technically, `char` isn't guaranteed to be `uint8_t`, so I'm not certain about the most portable way to fix this.  I suppose we could add the string lengths (hard-coded) into `static struct scrypt_test`, but that feels like a hack.

```
td@gin: ~/src/scrypt/test (compile-tests)
$ clang --version
Ubuntu clang version 3.4-1ubuntu3 (tags/RELEASE_34/final) (based on LLVM 3.4)
Target: x86_64-pc-linux-gnu
Thread model: posix
td@gin: ~/src/scrypt/test (compile-tests)
$ CC=clang bmake
clang -O2 -Wno-error -I .. -DCONFIG_H_FILE=\"config_freebsd.h\" -I ../libcperciva/alg -I ../libcperciva/cpusupport -I ../lib/crypto -I ../libcperciva/util -I . -D CPUSUPPORT_CONFIG_FILE=\"cpusupport-config.h\"  -Wall -Wstrict-prototypes -Wmissing-prototypes -Wpointer-arith -Wno-uninitialized -Wreturn-type -Wcast-qual -Wpointer-arith -Wwrite-strings -Wswitch -Wshadow -Werror   -c main.c
main.c:35:17: error: passing 'const char *' to parameter of type
      'const uint8_t *' (aka 'const unsigned char *') converts between pointers
      to integer types with different sign [-Werror,-Wpointer-sign]
                crypto_scrypt(test->passwd, strlen(test->passwd),
                              ^~~~~~~~~~~~
../lib/crypto/crypto_scrypt.h:44:34: note: passing argument to parameter here
int crypto_scrypt(const uint8_t *, size_t, const uint8_t *, size_t, uint64_t,
                                 ^
main.c:36:7: error: passing 'const char *' to parameter of type
      'const uint8_t *' (aka 'const unsigned char *') converts between pointers
      to integer types with different sign [-Werror,-Wpointer-sign]
                    test->salt, strlen(test->salt),
                    ^~~~~~~~~~
../lib/crypto/crypto_scrypt.h:44:59: note: passing argument to parameter here
int crypto_scrypt(const uint8_t *, size_t, const uint8_t *, size_t, uint64_t,
                                                          ^
main.c:37:34: error: passing 'char [64]' to parameter of type 'uint8_t *'
      (aka 'unsigned char *') converts between pointers to integer types with
      different sign [-Werror,-Wpointer-sign]
                    test->N, test->r, test->p, kbuf, 64);
                                               ^~~~
../lib/crypto/crypto_scrypt.h:45:34: note: passing argument to parameter here
    uint32_t, uint32_t, uint8_t *, size_t);
                                 ^
3 errors generated.
*** Error code 1

Stop.
bmake: stopped in /home/td/src/scrypt/test
```

**Comment by cperciva** [comment]
In FreeBSD we use -Wno-pointer-sign to avoid this sort of silliness.  In this case it's perfectly safe to cast between 'char *' and 'uint8_t *'; if you want to silence the warnings, go ahead and add casts.

**Comment by gperciva** [comment]
where on earth does `bmake` get its `CFLAGS` from?!  I don't want `-Werror`!  (actually, I was trying to set `-Wno-cast-qual` in the `Makefile` but it wasn't getting picked up.

```
td@gin: ~/src/scrypt (compile-tests)
$ grep "Werror" * -R
td@gin: ~/src/scrypt (compile-tests)
$ cd test/
td@gin: ~/src/scrypt/test (compile-tests)
$ bmake
cc -O2 -Wno-error -I .. -DCONFIG_H_FILE=\"config_freebsd.h\" -I ../libcperciva/alg -I ../libcperciva/cpusupport -I ../lib/crypto -I ../libcperciva/util -I . -D CPUSUPPORT_CONFIG_FILE=\"cpusupport-config.h\"  -Wall -Wstrict-prototypes -Wmissing-prototypes -Wpointer-arith -Wno-uninitialized -Wreturn-type -Wcast-qual -Wpointer-arith -Wwrite-strings -Wswitch -Wshadow -Werror   -c main.c
main.c: In function ‘main’:
main.c:35:17: error: cast discards ‘__attribute__((const))’ qualifier from pointer target type [-Werror=cast-qual]
   crypto_scrypt((uint8_t*)test->passwd, strlen(test->passwd),
                 ^
main.c:36:7: error: cast discards ‘__attribute__((const))’ qualifier from pointer target type [-Werror=cast-qual]
       (uint8_t*)test->salt, strlen(test->salt),
       ^
cc1: all warnings being treated as errors
*** Error code 1

Stop.
bmake: stopped in /home/td/src/scrypt/test
td@gin: ~/src/scrypt/test (compile-tests)
$ echo $CFLAGS

```

**Comment by gperciva** [comment]
ok, found it.  `/usr/share/mk/bsd.sys.mk`, apparently.

**Comment by cperciva** [comment]
Right.  Which is included indirectly from `<bsd.prog.mk>`.  With different CFLAGS set based on the OS, version, and the value of WARNS used.

**Comment by cperciva** [comment]
That last warning was correct though, you need to cast to `(const uint8_t *)`.

**Comment by gperciva** [comment]
but why are those flags put at the *end* of the compile command?  I mean, I can't override them with CFLAGS, and isn't that the whole point of CFLAGS?

**Comment by cperciva** [comment]
I don't know the exact history of the FreeBSD bsd make logic, and you're probably using NetBSD's version anyway.  I tend to set WARNS=0 if I want to silence warnings temporarily.

**Comment by gperciva** [comment]
This was fixed in #13.
